src/Raven.Server/Documents/PeriodicBackup/PeriodicBackupRunner.cs

```diff
@@ -0,0 +1,934 @@
+            var isFullBackup = nextFullBackup != null && 
+                (nextIncrementalBackup == null || nextFullBackup <= nextIncrementalBackup);
+
+            var nextBackupDateTime = isFullBackup ? nextFullBackup.Value : nextIncrementalBackup.Value;
+            var nextBackupTimeSpan =
+                (nextBackupDateTime - now).Ticks <= 0 ?
+                    TimeSpan.Zero :
+                    nextBackupDateTime - now;
+
+            if (_logger.IsInfoEnabled)
+                _logger.Info($"Next {(isFullBackup ? "full" : "incremental")} " +
+                             $"backup is in {nextBackupTimeSpan.TotalMinutes} minutes");
+
+            return GenerateNewTimer(new BackupTaskDetails
+            {
+                IsFullBackup = isFullBackup,
+                TaskId = configuration.TaskId.Value,
+                NextBackup = nextBackupTimeSpan
+            });
+        }
+
+        private static DateTime? GetNextBackupOccurrence(string backupFrequency, DateTime now)
+        {
+            try
+            {
+                var backupParser = CrontabSchedule.Parse(backupFrequency);
+                return backupParser.GetNextOccurrence(now);
+            }
+            catch (Exception e)
+            {
+                var exception = e;
+                //TODO: error to notification center
+                return null;
+            }
+        }
+
+        private class BackupTaskDetails
+        {
+            public long TaskId { get; set; }
+
+            public bool IsFullBackup { get; set; }
+
+            public TimeSpan NextBackup { get; set; }
+        }
+
+        private Timer GenerateNewTimer(BackupTaskDetails backupTaskDetails)
+        {
+            var isValidTimeSpanForTimer = IsValidTimeSpanForTimer(backupTaskDetails.NextBackup);
+            var timer = isValidTimeSpanForTimer ? 
+                new Timer(TimerCallback, backupTaskDetails, backupTaskDetails.NextBackup, Timeout.InfiniteTimeSpan) : 
+                new Timer(LongPeriodTimerCallback, backupTaskDetails, MaxTimerTimeout, Timeout.InfiniteTimeSpan);
+
+            return timer;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private bool IsValidTimeSpanForTimer(TimeSpan nextBackupTimeSpan)
+        {
+            return nextBackupTimeSpan < MaxTimerTimeout;
+        }
+
+        private void TimerCallback(object backupTaskDetails)
+        {
+            if (_cancellationToken.IsCancellationRequested)
+                return;
+
+            var backupDetails = (BackupTaskDetails)backupTaskDetails;
+
+            PeriodicBackup periodicBackup;
+            if (ShouldRunBackupAfterTimerCallback(backupDetails, out periodicBackup) == false)
+                return;
+
+            CreateBackupTask(periodicBackup, backupDetails);
+        }
+
+        private void CreateBackupTask(PeriodicBackup periodicBackup, BackupTaskDetails backupDetails)
+        {
+            periodicBackup.RunningTask = Task.Run(async () =>
```

Comment:
There can't be concurrent runs here.
Only one backup task is running at any given point. After it's finished, we create a timer that will run only once and so on...